ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java

```diff
@@ -197,4 +225,313 @@ public void updateAccounts(Collection<Account> accounts, int batchSize) throws S
+    try (ResultSet rs = getSinceStatement.executeQuery()) {
+      List<Container> containers = convertContainersResultSet(rs);
+      dataAccessor.onSuccess(Read, System.currentTimeMillis() - startTimeMs);
+      return containers;
+    } catch (SQLException e) {
+      dataAccessor.onException(e, Read);
+      throw e;
+    }
+  }
+
+  /**
+   * Convert a query result set to a list of containers.
+   * @param resultSet the result set.
+   * @return a list of {@link Container}s.
+   * @throws SQLException
+   */
+  private List<Container> convertContainersResultSet(ResultSet resultSet) throws SQLException {
+    List<Container> containers = new ArrayList<>();
+    while (resultSet.next()) {
+      int accountId = resultSet.getInt(ACCOUNT_ID);
+      String containerJson = resultSet.getString(CONTAINER_INFO);
+      Timestamp lastModifiedTime = resultSet.getTimestamp(LAST_MODIFIED_TIME);
+      int version = resultSet.getInt(VERSION);
+      Container container = new ContainerBuilder(
+          Container.fromJson(new JSONObject(containerJson), (short) accountId)).setLastModifiedTime(
+          lastModifiedTime.getTime()).setSnapshotVersion(version).build();
+      containers.add(container);
+    }
+    return containers;
+  }
+
+  /**
+   * Adds/Updates accounts and their containers to the database in batches atomically using transaction.
+   * @param accountsInfo information of updated Accounts
+   * @param batchSize number of statements to be executed in one batch
+   * @throws SQLException
+   */
+  public void updateAccountsAndContainers(List<AccountUtils.AccountUpdateInfo> accountsInfo, int batchSize)
+      throws SQLException {
+    Connection conn = dataAccessor.getDatabaseConnection(true);
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      PreparedStatement insertAccountStatement = dataAccessor.getPreparedStatement(insertAccountsSql, true);
+      PreparedStatement updateAccountStatement = dataAccessor.getPreparedStatement(updateAccountsSql, true);
+      PreparedStatement insertContainerStatement = dataAccessor.getPreparedStatement(insertContainersSql, true);
+      PreparedStatement updateContainerStatement = dataAccessor.getPreparedStatement(updateContainersSql, true);
+
+      conn.setAutoCommit(false);
+      int batchCount = 0, insertAccountCount = 0, updateAccountCount = 0, insertContainerCount = 0,
+          updateContainerCount = 0;
+      for (int i = 0; i < accountsInfo.size(); i++) {
+        Account account = accountsInfo.get(i).getAccount();
+        boolean isAccountModified = accountsInfo.get(i).isModified();
+        List<Container> addedContainers = accountsInfo.get(i).getAddedContainers();
+        List<Container> updatedContainers = accountsInfo.get(i).getModifiedContainers();
+
+        if (isAccountModified) {
+          if (account.getSnapshotVersion() == 0) {
+            insertAccountStatement.setString(1, AccountCollectionSerde.accountToJsonNoContainers(account).toString());
```

Comment:
Why are we doing an insert if isAccountModified is true?  Maybe I don't understand what that variable means.